Remove HTML conformance concepts from MathML code

annevk opened the pull request:
#### 3719e049046bfe237cc3d0aa041735cbb8a68d41
<pre>
Remove HTML conformance concepts from MathML code
<a href="https://bugs.webkit.org/show_bug.cgi?id=256833">https://bugs.webkit.org/show_bug.cgi?id=256833</a>
rdar://109710721

Reviewed by Frédéric Wang.

MathML should not be itself validating what elements can and cannot be
nested inside of it. It&apos;s really a function of the node tree and CSS
whether something ends up rendering or not.

* LayoutTests/imported/w3c/web-platform-tests/html/syntax/parsing/html-integration-point-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/syntax/parsing/math-parse03-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/inert/inert-on-non-html-expected.txt:
* LayoutTests/mathml/presentation/foreign-element-in-annotation-xml-expected.txt:
* LayoutTests/mathml/presentation/foreign-element-in-annotation-xml.html:
* LayoutTests/mathml/presentation/foreign-element-in-token-expected.txt:
* LayoutTests/mathml/presentation/foreign-element-in-token.html:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-childlist-002-expected.txt: Renamed from LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-childlist-002-expected.txt.
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-childlist-002-expected.txt: Removed.
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-childlist-002-expected.txt:
* LayoutTests/platform/mac-ventura/imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-childlist-002-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-childlist-002-expected.txt: Renamed from LayoutTests/platform/mac-wk2/imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-childlist-002-expected.txt.
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-childlist-002-expected.txt: Removed.
* Source/WebCore/mathml/MathMLAnnotationElement.cpp:
(WebCore::MathMLAnnotationElement::childShouldCreateRenderer const):
* Source/WebCore/mathml/MathMLPresentationElement.cpp:
(WebCore::MathMLPresentationElement::isPhrasingContent): Deleted.
(WebCore::MathMLPresentationElement::isFlowContent): Deleted.
* Source/WebCore/mathml/MathMLPresentationElement.h:
* Source/WebCore/mathml/MathMLTokenElement.cpp:
(WebCore::MathMLTokenElement::childShouldCreateRenderer const):

Canonical link: <a href="https://commits.webkit.org/264657@main">https://commits.webkit.org/264657@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9e4171923000069a65d4389e6984ac51563dfe6f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8232 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8524 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8741 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9899 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8298 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/8241 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10512 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8435 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11173 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8377 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9443 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/7466 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/10044 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6742 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7531 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/15091 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/7865 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7662 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/11021 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8137 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6617 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7427 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/7443 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1988 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11635 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7878 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fred-wang commented on Source/WebCore/mathml/MathMLSelectElement.cpp:
```diff
@@ -187,10 +187,7 @@ Element* MathMLSelectElement::getSelectedSemanticsChild()
-            // If the <annotation-xml> element has an encoding attribute describing presentation MathML, SVG or HTML we assume the content can be displayed and we stop here.
-            auto& value = child->attributeWithoutSynchronization(MathMLNames::encodingAttr);
-            if (isMathMLEncoding(value) || isSVGEncoding(value) || isHTMLEncoding(value))
-                return child;
```
This is to implement the legacy algorithm described on https://developer.mozilla.org/en-US/docs/Web/MathML/Element/semantics ; ideally we should remove this when the MathMLCore flag is on (or if Apple is OK with dropping that implementation) but let's keep this for now.

annevk replied:
The value of this attribute doesn't strike me as something that should matter for rendering purposes (although unfortunately it does mean something for the HTML parser), but we can keep it around for now, sure.

fred-wang replied:
I mean, let's keep this in the present PR, and see if we can remove the implementation mentioned in the note of https://developer.mozilla.org/en-US/docs/Web/MathML/Element/semantics in follow-up PRs.

annevk replied:
I'm sorry, I'm still a little lost.

This PR is no longer making those changes to `MathMLSelectElement`. Do you want more changes taken out?

It's not clear to me what MathMLCore is or how an MDN reference is supposed to help me. I scanned through https://w3c.github.io/mathml/ and could not find any relevant requirements surrounding `encoding`, other than a bunch of stuff for people writing MathML (and some contradictions such as it apparently containing a MIME type but then not really as `MathML-Content` is not a MIME type).

fred-wang replied:
TL;DR: The PR looks good currently!

MathML Core is the spec that is intended to be implemented in browsers: https://w3c.github.io/mathml-core ; the implementation of `<semantics>` there is just "render the first child":

```
semantics > :not(:first-child) {
  display: none;
}
```

The MathML spec at https://w3c.github.io/mathml is not always great for implementation purpose (lacks a lot of details, tests...) but some of the features mentioned there are still implemented in WebKit/Firefox. In particular WebKit still implements the case "when a presentation MathML annotation is present in a semantics element, it may be used as the default rendering of the semantics element, instead of the default rendering of the first child" (https://w3c.github.io/mathml/#mixing_alternate_representations) and more generally the behavior described in the MDN note (but not in any spec) "Legacy MathML specifications allowed renderers to decide the default rendering according to available annotations" (https://developer.mozilla.org/en-US/docs/Web/MathML/Element/semantics).

I believe ideally we just want to simplify the implementation and align it with what is described MathML Core (in particular ignore `encoding` for rendering purpose). But in the past some people at Apple were against unshipping features, that's why I suggested you not to do it in that PR.

fred-wang commented on Source/WebCore/mathml/MathMLTokenElement.cpp:
```diff
@@ -74,10 +74,10 @@ RenderPtr<RenderElement> MathMLTokenElement::createElementRenderer(RenderStyle&&
+bool MathMLTokenElement::childShouldCreateRenderer(const Node&) const
 {
     // The HTML specification defines <mi>, <mo>, <mn>, <ms> and <mtext> as insertion points.
-    return isPhrasingContent(child) && StyledElement::childShouldCreateRenderer(child);
```
Similarly can we just defer to the parent of MathMLElement here:   `StyledElement::childShouldCreateRenderer(child);`

fred-wang commented on Source/WebCore/mathml/MathMLAnnotationElement.cpp:
```diff
@@ -71,35 +71,13 @@ bool MathMLAnnotationElement::childShouldCreateRenderer(const Node& child) const
 
 void MathMLAnnotationElement::attributeChanged(const QualifiedName& name, const AtomString& oldValue, const AtomString& newValue, AttributeModificationReason reason)
 {
-    if (name == MathMLNames::srcAttr || name == MathMLNames::encodingAttr) {
```
This is still needed (see my comment in MathMLSelectElement::getSelectedSemanticsChild below)

fred-wang commented on Source/WebCore/mathml/MathMLAnnotationElement.cpp:
```diff
@@ -71,35 +71,13 @@ bool MathMLAnnotationElement::childShouldCreateRenderer(const Node& child) const
-        return is<HTMLHtmlElement>(htmlElement) || (isFlowContent(htmlElement) && StyledElement::childShouldCreateRenderer(child));
-    }
-
-    return false;
```
Can we just defer to the parent class of MathMLElement:

`return StyledElement::childShouldCreateRenderer(child);`

? Checking a bit, this will call Element::childShouldCreateRenderer where the difference is minimal (for SVG child only accept valid `<svg>`).

fred-wang commented on LayoutTests/mathml/presentation/foreign-element-in-annotation-xml-expected.txt:
```diff
@@ -35,9 +35,10 @@ h6
 
 mathml
 
-
```
Can we update `foreign-element-in-annotation-xml.html` to use something else than "ERROR"? The comments should also be updated since `<annotation-xml>` allows more children to create renderers.

Ideally the test should be extended too but maybe the HTML parser will forbid some constructions and one will have to create relevant trees using DOM operations instead...

fred-wang commented on LayoutTests/mathml/presentation/foreign-element-in-token-expected.txt:
```diff
@@ -455,18 +455,73 @@ mo
 Foreign non-phrasing content should not create renderers:
```
Can we update the text in foreign-element-in-token.html now that foreign non-phrasing content do create renderers?

fred-wang reviewed (approved):
Thanks for working on this Anne. LGTM with the comments below addressed.

annevk committed:
Remove HTML conformance concepts from MathML code
https://bugs.webkit.org/show_bug.cgi?id=256833
rdar://109710721

Reviewed by Frédéric Wang.

MathML should not be itself validating what elements can and cannot be
nested inside of it. It's really a function of the node tree and CSS
whether something ends up rendering or not.

* LayoutTests/imported/w3c/web-platform-tests/html/syntax/parsing/html-integration-point-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/syntax/parsing/math-parse03-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/inert/inert-on-non-html-expected.txt:
* LayoutTests/mathml/presentation/foreign-element-in-annotation-xml-expected.txt:
* LayoutTests/mathml/presentation/foreign-element-in-annotation-xml.html:
* LayoutTests/mathml/presentation/foreign-element-in-token-expected.txt:
* LayoutTests/mathml/presentation/foreign-element-in-token.html:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-childlist-002-expected.txt: Renamed from LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-childlist-002-expected.txt.
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-childlist-002-expected.txt: Removed.
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-childlist-002-expected.txt:
* LayoutTests/platform/mac-ventura/imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-childlist-002-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-childlist-002-expected.txt: Renamed from LayoutTests/platform/mac-wk2/imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-childlist-002-expected.txt.
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-childlist-002-expected.txt: Removed.
* Source/WebCore/mathml/MathMLAnnotationElement.cpp:
(WebCore::MathMLAnnotationElement::childShouldCreateRenderer const):
* Source/WebCore/mathml/MathMLPresentationElement.cpp:
(WebCore::MathMLPresentationElement::isPhrasingContent): Deleted.
(WebCore::MathMLPresentationElement::isFlowContent): Deleted.
* Source/WebCore/mathml/MathMLPresentationElement.h:
* Source/WebCore/mathml/MathMLTokenElement.cpp:
(WebCore::MathMLTokenElement::childShouldCreateRenderer const):

Canonical link: https://commits.webkit.org/264657@main
